cpp/src/arrow/array/array_nested.cc

```diff
@@ -216,12 +216,26 @@ static std::shared_ptr<Array> SliceArrayWithOffsets(const Array& array, int64_t
   return array.Slice(begin, end - begin);
 }
 
+namespace {
+struct FlattenWithRecursion {
+  // Flatten all list-like types array recursively
+  static Result<std::shared_ptr<Array>> Flatten(const Array& array, bool with_recursion,
+                                                MemoryPool* memory_pool);
+};
+}  // namespace
```

Comment:
I'm suggesting `LogicalList` instead of `ListLike` because the `is_list_like` type predicate (that we can't change for backwards compatibility reasons) doesn't include list-views.